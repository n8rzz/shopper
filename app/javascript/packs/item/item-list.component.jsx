import React from 'react';
import PropTypes from 'prop-types';
import { Item } from './item.component';

export function ItemList(props) {
    if (props.itemList.length === 0) {
        return null;
    }

    return (
        <ul className={'vlist'}>
            {props.itemList.map((item, index) => (
                <Item
                    csrf={props.csrf}
                    departmentId={item.department_id}
                    departmentName={props.departmentMap[item.department_id].name}
                    itemId={item.id}
                    itemName={item.name}
                    key={`item-${index}`}
                />
            ))}
        </ul>
    );
}

ItemList.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentMap: PropTypes.object.isRequired,
    itemList: PropTypes.array.isRequired,
};
