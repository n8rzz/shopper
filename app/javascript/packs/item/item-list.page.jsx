import React from 'react';
import PropTypes from 'prop-types';
import _groupBy from 'lodash/groupBy';
import _sortBy from 'lodash/sortBy';
import {FilterBar} from '../components/filter-bar/filter-bar.component';
import {ItemList} from './item-list.component';
import { GroupedItemList } from './grouped-item-list.component';

// TODO: abstract these constants to another file
export const FILTER_CONCERN = {
    ALPHA: 'ALPHA',
    DEPARTMENT: 'DEPARTMENT',
    FILTER: 'FILTER',
};
export const FILTER_LABEL = {
    [FILTER_CONCERN.ALPHA]: 'Alpha',
    [FILTER_CONCERN.DEPARTMENT]: 'Department',
    [FILTER_CONCERN.FILTER]: 'Filter',
}
export const FILTER_BAR_ITEMS = [
    {
        text: FILTER_LABEL.ALPHA,
        url: '',
    },
    {
        text: FILTER_LABEL.DEPARTMENT,
        url: '',
    },
    {
        text: FILTER_LABEL.FILTER,
        url: '',
    },
];
export const DEFAULT_DRILL_DOWN_VALUE = '--';

export class ItemListPage extends React.Component {
    constructor(props) {
        super(props);

        this._departmentMap = this._buildDepartmentMap(props);
        this._itemsByFirstLetter = this._buildItemsByFirstLetter(props);
        this._filterableItemList = Object.keys(this._itemsByFirstLetter);
        this.state = {
            concern: FILTER_CONCERN.ALPHA,
            itemsByFilter: props.items,
        };
    }

    componentWillMount() {
        this._onChangeFilterHandler = this._onChangeFilter.bind(this);
        this._onChangeLetterFilterValueHandler = this._onChangeLetterFilterValue.bind(this);
    }

    componentWillUnmount() {
        this._onChangeFilterHandler = null;
        this._onChangeLetterFilterValueHandler = null;
    }

    _buildSecondaryFilterOptionListJsx() {
        return (
            <select
                className="input-select"
                onChange={this._onChangeLetterFilterValueHandler}
            >
                <option value={DEFAULT_DRILL_DOWN_VALUE}>{DEFAULT_DRILL_DOWN_VALUE}</option>
                {this._filterableItemList.map((filterableLetter, index) => {
                    return (
                        <option value={filterableLetter} key={`select-${index}`}>
                            {filterableLetter.toUpperCase()}
                        </option>
                    );
                })}
            </select>
        );
    }

    _buildItemListForActiveConcernJsx() {
        switch (this.state.concern) {
            case FILTER_CONCERN.DEPARTMENT:
                return (
                    <GroupedItemList
                        csrf={this.props.csrf}
                        departments={this.props.departments}
                        departmentMap={this._departmentMap}
                        items={this.props.items}
                    />
                );
            case FILTER_CONCERN.FILTER:
                return (
                    <React.Fragment>
                        <div className="vr_3">
                            {this._buildSecondaryFilterOptionListJsx()}
                        </div>

                        <ItemList
                            csrf={this.props.csrf}
                            departmentMap={this._departmentMap}
                            itemList={this.state.itemsByFilter}
                        />
                    </React.Fragment>
                );
            default:
                return (
                    <ItemList
                        csrf={this.props.csrf}
                        departmentMap={this._departmentMap}
                        itemList={this.props.items}
                    />
                );
        }
    }

    render() {
        return (
            <React.Fragment>
                <FilterBar
                    items={FILTER_BAR_ITEMS}
                    onChangeFilterHandler={this._onChangeFilterHandler}
                />

                {this._buildItemListForActiveConcernJsx()}
            </React.Fragment>
        );
    }

    // TODO: rename this
    _onChangeLetterFilterValue(event) {
        const nextLetterFilter = event.target.value;
        let itemsByFilter = this._itemsByFirstLetter[nextLetterFilter];

        if (nextLetterFilter === DEFAULT_DRILL_DOWN_VALUE) {
            itemsByFilter = this.props.items;
        }

        this.setState({ itemsByFilter });
    }

    _onChangeFilter(event) {
        event.preventDefault();

        const filterName = event.currentTarget.text.toUpperCase();

        this.setState({ concern: FILTER_CONCERN[filterName] });
    }

    _buildDepartmentMap(props) {
        return props.departments.reduce((sum, department) => {
            sum[department.id] = department;

            return sum;
        }, {});
    }

    _buildItemsByFirstLetter(props) {
        return props.items.reduce((sum, item) => {
            const firstLetter = item.name.substr(0, 1).toLowerCase();

            if (typeof sum[firstLetter] !== 'undefined') {
                sum[firstLetter] = [
                    ...sum[firstLetter],
                    item,
                ];

                return sum;
            }

            sum[firstLetter] = [item];

            return sum;
        }, {});
    }
}

ItemListPage.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array,
    items: PropTypes.array,
};
