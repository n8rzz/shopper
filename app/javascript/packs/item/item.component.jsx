import React from 'react';
import PropTypes from 'prop-types';
import { StepperContainer } from '../components/stepper/stepper.container';

export function Item(props) {
    const editItemPath = `items/${props.itemId}/edit`;

    return (
        <li className={'js-item'}>
            <div className={'listItem mix-listItem_flex'}>
                <div className={'listItem-content'}>
                    <div className={'listItem-hd'}>
                        <h2 className={'hdg hdg_2'}>
                            <a href={editItemPath}>
                                {props.itemName}
                            </a>
                        </h2>
                    </div>
                    <div className={'listItem-ft'}>
                        { props.departmentName }
                    </div>
                </div>
                <div className={'listItem-action'}>
                    <StepperContainer
                        csrf={props.csrf}
                        departmentId={props.departmentId}
                        itemId={props.itemId}
                    />
                </div>
            </div>
        </li>
    );
}

Item.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentId: PropTypes.number.isRequired,
    departmentName: PropTypes.string.isRequired,
    itemId: PropTypes.number.isRequired,
    itemName: PropTypes.string.isRequired,
};
