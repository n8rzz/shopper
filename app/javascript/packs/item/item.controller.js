import FlashController from '../flash/flash.controller';
import OrderItemModel from './order-item.model';
import ApiService from '../service/api.service';

const CLASSNAMES = {
    ROOT_ELEMENT: 'js-item',
    ACTION_ITEM: 'js-item-action',
    ADD_ITEM_TO_ORDER_FORM: 'js-item-createOrderItemForm'
};

export class OrderController {
    constructor() {
        const $element = document.getElementsByClassName(CLASSNAMES.ROOT_ELEMENT);

        if ($element.length === 0) {
            this.disable();

            return;
        }

        this._$actionItemElementList = null;
        this._flashController = new FlashController();
        this._isEnabled = false;

        return this._init()
            ._createChildren()
            ._setupHandlers()
            .enable();
    }

    _init() {
        return this;
    }

    _createChildren() {
        this._$actionItemElementList = document.getElementsByClassName(CLASSNAMES.ACTION_ITEM);

        return this;
    }

    _setupHandlers() {
        this._onClickAddItemToOrderHandler = this._onClickAddItemToOrder.bind(this);

        return this;
    }

    enable() {
        if (this._isEnabled) {
            return this;
        }

        this._addHandlerToElementList(this._$actionItemElementList, this._onClickAddItemToOrderHandler);

        this._isEnabled = true;

        return this;
    }

    disable() {
        if (!this._isEnabled) {
            return this;
        }

        this._isEnabled = false;

        return this;
    }

    _addHandlerToElementList(elementList, handler) {
        if (elementList.length === 0) {
            return;
        }

        for (let i =0; i < elementList.length; i++) {
            const element = elementList[i];

            element.addEventListener('click', handler);
        }
    }

    _onClickAddItemToOrder(event) {
        event.stopPropagation();
        event.preventDefault();

        const csrf = event.currentTarget.dataset.csrf;
        const createOrderItemUrl = `/order_items/create/item.json`;
        const orderItemModel = new OrderItemModel(event.currentTarget);

        if (!orderItemModel.isValid()) {
            return;
        }

        ApiService.post(createOrderItemUrl, orderItemModel.toJson(), csrf)
            .then((response) => {
                if (response.status >= 300) {
                    console.error(`Received an unexpected status code from ${createOrderItemUrl}. See response: ${response}`);

                    return;
                }

                this._flashController.showNoticeMessage(`${orderItemModel.name} added to pending order`);
            })
            .catch((error) => {
                // TODO: add ability to set flash warning, error
                this._flashController.showNoticeMessage('Something went wrong, Item was not successfully added to Order');

                throw error;
            });
    }
}

document.addEventListener('turbolinks:load', () => {
    const singleOrderController = new OrderController();
});
