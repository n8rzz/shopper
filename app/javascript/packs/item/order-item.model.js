/**
 * Accepts a DOM element (`currentTarget` from a click event) and
 * extracts information from that DOM element used in submission
 * of a form.
 *
 * This model acts as a flyweight and should only be available
 * between mouse click and api request.
 *
 * @class OrderItemModel
 */
export default class OrderItemModel {
    constructor(element) {
        this.itemId = -1;
        this.name = '';
        this.departmentId = -1;
        this.qty = -1;

        return this._init(element);
    }

    isValid() {
        return this.itemId !== -1
            && this.departmentId !== -1
            && this.qty !== -1
            && this.qty > 0;
    }

    toJson() {
        return {
            item_id: this.itemId,
            department_id: this.departmentId,
            qty: this.qty,
        };
    }

    _init(element) {
        this.itemId = parseInt(element.dataset.itemId, 10);
        this.name = element.dataset.itemName;
        this.departmentId = parseInt(element.dataset.departmentId, 10);
        this.qty = this._extractQtyFromInputElement();

        return this;
    }

    _extractQtyFromInputElement() {
        const qtyElement = document.getElementById(`item-${this.itemId}-qty`);

        return parseInt(qtyElement.value, 10);
    }
}
