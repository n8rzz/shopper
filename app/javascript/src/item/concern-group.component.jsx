import React from 'react';
import PropTypes from 'prop-types';
import { ItemList } from './item-list.component';

export function ConcernGroup(props) {
    return (
        <div
            className={'concernGroup'}
            key={`concernGroup-${props.groupName}`}
        >
            <div className={'concernGroup-hd'}>
                <h2>{props.groupName}</h2>
            </div>
            <ItemList
                csrf={props.csrf}
                departmentMap={props.departmentMap}
                itemList={props.itemList}
            />
        </div>
    );
}

ConcernGroup.propTypes = {
    csrf: PropTypes.string.isRequired,
    departmentMap: PropTypes.object.isRequired,
    groupName: PropTypes.string.isRequired,
    itemList: PropTypes.array.isRequired,
};
