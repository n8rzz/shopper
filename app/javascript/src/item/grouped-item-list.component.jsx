import React from 'react';
import PropTypes from 'prop-types';
import _groupBy from 'lodash/groupBy';
import _sortBy from 'lodash/sortBy';
import { ConcernGroup } from './concern-group.component';

export class GroupedItemList extends React.PureComponent {
    constructor(props) {
        super(props);

        this._sortedDepartmentList = _sortBy(this.props.departments, 'name');
        this._groupedItemList = _groupBy(this.props.items, 'department_id');
    }

    render() {
        const groupListJsx = this._sortedDepartmentList.map((department, index) => {
            const itemsForDepartmentId = this._groupedItemList[department.id];

            if (typeof itemsForDepartmentId === 'undefined') {
                return null;
            }

            return (
                <ConcernGroup
                    csrf={this.props.csrf}
                    departmentMap={this.props.departmentMap}
                    groupName={department.name}
                    key={`ConcernGroup-${index}`}
                    itemList={itemsForDepartmentId}
                />
            );
        });

        return (
            <React.Fragment>
                {groupListJsx}
            </React.Fragment>
        );
    }
}

GroupedItemList.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array.isRequired,
    departmentMap: PropTypes.object.isRequired,
    items: PropTypes.array.isRequired,
};
