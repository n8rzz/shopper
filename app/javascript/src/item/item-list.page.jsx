import React from 'react';
import PropTypes from 'prop-types';
import { FilterBar } from '../components/filter-bar/filter-bar.component';
import { ItemList } from './item-list.component';
import { GroupedItemList } from './grouped-item-list.component';
import {
    FILTER_CONCERN,
    FILTER_LABEL,
} from '../constants/filter-concern';

export const FILTER_BAR_ITEMS = [
    {
        text: FILTER_LABEL.ALPHA,
        url: '',
    },
    {
        text: FILTER_LABEL.DEPARTMENT,
        url: '',
    },
    {
        text: FILTER_LABEL.FILTER,
        url: '',
    },
];
export const DEFAULT_DRILL_DOWN_VALUE = '--';

export class ItemListPage extends React.Component {
    constructor(props) {
        super(props);

        this._departmentMap = this._buildDepartmentMap(props);
        this._itemsByFirstLetter = this._buildItemsByFirstLetter(props);
        this._filterableItemList = Object.keys(this._itemsByFirstLetter);
        this._onChangeFilterHandler = this._onChangeFilter.bind(this);
        this._onChangeLetterFilterValueHandler = this._onChangeLetterFilterValue.bind(this);
        this.state = {
            concern: FILTER_CONCERN.ALPHA,
            itemsByFilter: props.items,
        };
    }

    componentWillUnmount() {
        this._onChangeFilterHandler = null;
        this._onChangeLetterFilterValueHandler = null;
    }

    _buildSecondaryFilterOptionListJsx() {
        return (
            <select
                name={'filter-item-by-first-letter'}
                className={'input-select'}
                onChange={this._onChangeLetterFilterValueHandler}
            >
                <option value={DEFAULT_DRILL_DOWN_VALUE}>{DEFAULT_DRILL_DOWN_VALUE}</option>
                {this._filterableItemList.map((filterableLetter, index) => (
                    <option
                        value={filterableLetter}
                        key={`select-${index}`}
                    >
                        {filterableLetter.toUpperCase()}
                    </option>
                ))}
            </select>
        );
    }

    _buildItemListForActiveConcernJsx() {
        switch (this.state.concern) {
            case FILTER_CONCERN.DEPARTMENT:
                return (
                    <GroupedItemList
                        csrf={this.props.csrf}
                        departments={this.props.departments}
                        departmentMap={this._departmentMap}
                        items={this.props.items}
                    />
                );
            case FILTER_CONCERN.FILTER:
                return (
                    <React.Fragment>
                        <div className={'vr_3'}>
                            {this._buildSecondaryFilterOptionListJsx()}
                        </div>

                        <ItemList
                            csrf={this.props.csrf}
                            departmentMap={this._departmentMap}
                            itemList={this.state.itemsByFilter}
                        />
                    </React.Fragment>
                );
            default:
                return (
                    <ItemList
                        csrf={this.props.csrf}
                        departmentMap={this._departmentMap}
                        itemList={this.props.items}
                    />
                );
        }
    }

    // TODO: rename this
    _onChangeLetterFilterValue(event) {
        const nextLetterFilter = event.target.value;
        let itemsByFilter = this._itemsByFirstLetter[nextLetterFilter];

        if (nextLetterFilter === DEFAULT_DRILL_DOWN_VALUE) {
            itemsByFilter = this.props.items;
        }

        this.setState({ itemsByFilter });
    }

    _onChangeFilter(event) {
        event.preventDefault();

        const filterName = event.currentTarget.text.toUpperCase();

        this.setState({ concern: FILTER_CONCERN[filterName] });
    }

    _buildDepartmentMap(props) {
        return props.departments.reduce((sum, department) => {
            sum[department.id] = department;

            return sum;
        }, {});
    }

    _buildItemsByFirstLetter(props) {
        return props.items.reduce((sum, item) => {
            const firstLetter = item.name.substr(0, 1).toLowerCase();

            if (typeof sum[firstLetter] !== 'undefined') {
                sum[firstLetter] = [
                    ...sum[firstLetter],
                    item,
                ];

                return sum;
            }

            sum[firstLetter] = [item];

            return sum;
        }, {});
    }

    render() {
        return (
            <React.Fragment>
                <FilterBar
                    initialSelection={FILTER_LABEL.ALPHA}
                    items={FILTER_BAR_ITEMS}
                    onChangeFilterHandler={this._onChangeFilterHandler}
                />
                {this._buildItemListForActiveConcernJsx()}
            </React.Fragment>
        );
    }
}

ItemListPage.propTypes = {
    csrf: PropTypes.string.isRequired,
    departments: PropTypes.array.isRequired,
    items: PropTypes.array.isRequired,
};
