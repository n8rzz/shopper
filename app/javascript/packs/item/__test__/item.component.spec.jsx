import React from 'react';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import { Item } from '../item.component';
import { csrfMock } from '../../__mocks__/csrf.mock';
import { itemMock } from '../../__mocks__/item.mock';

describe('Item', () => {
    test('renders correctly with valid props', () => {
        const component = shallow(<Item
            csrf={csrfMock}
            departmentId={itemMock.department_id}
            departmentName={'meat'}
            itemId={itemMock.id}
            itemName={itemMock.name}
            isSubmitting={false}
            isSubmitSuccess={false}
            onSubmitHandler={sinon.spy()}
        />);

        expect(component).toMatchSnapshot();
    });
});
