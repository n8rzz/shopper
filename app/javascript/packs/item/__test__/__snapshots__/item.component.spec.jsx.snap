// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders correctly with valid props 1`] = `
<li
  className="js-item"
>
  <div
    className="listItem mix-listItem_flex"
  >
    <div
      className="listItem-content"
    >
      <div
        className="listItem-hd"
      >
        <h2
          className="hdg hdg_2"
        >
          <a
            className="link"
            href="items/11/edit"
          >
            Red Face Tuna
          </a>
        </h2>
      </div>
      <div
        className="listItem-ft"
      >
        meat
      </div>
    </div>
    <div
      className="listItem-action"
    >
      <StepperContainer
        csrf="3yIq75tyV791TFkzSZo/S4b+TVa/N5J3dI14PzMUPviGViEGeDSnGnvNix+MaB0RB0XuV7bpX8I3SmIEAr4/zQ=="
        departmentId={32}
        itemId={11}
        itemName="Red Face Tuna"
      />
    </div>
  </div>
</li>
`;
